Fix autoscaling and related issues

bandesz opened the pull request:
## Summary

This PR fixes various issues around autoscaling, Helm Applications and cluster components.

**Which issue(s) this PR resolves**:

Resolves #1019 - Do not install the EKS autoscaler in the Kore cluster
Resolves #974 - Cluster services are not installed in the correct order
Resolves #965 - Services are not updated when the configuration changes
Resolves #945 - Update Cluster service components on change
Resolves #967 - Services wrongly report successful applications with applications.app.k8s.io
Resolves #971 - Secret reference errors after deleting EKS cluster which has been auto-scaled
Resolves #983 - Services Helm App Delay

## Changes

 * **Do not install the EKS autoscaler in the Kore cluster** - we've accidentally installed this as all app plans were converted into an admin service when the application provider is reconciled. Now we only install the two apps explicitly.

 * **Save the autoscaling group names on the EKS Node Group** - this saves us some AWS API queries when setting up autoscaling

 * **Create the EKS Cluster OIDC provider on the EKS object and not in the autoscaler component** - the OIDC provider will be needed for any service accounts using IAM roles, so I moved the management to the EKS controller. It's always created.

 * **Move the EKS autoscaler to the cluster as a service, simplify cluster autoscaler IAM role** - this solves the wrong service install order + we only create a single embedded IAM policy on the IAM role instead of polluting the global policies. I give the IAM role access to all nodegroups in the cluster which simplifies the role but opens up the permissions a little.

 * **Fix autoscaler app resource selector as currently it doesn't match any resources** - I had to remove a label filter

 * **Update k8s-objectmatcher to 1.3.3 which sorts the JSON map keys and stops infinite reconcile loops caused by undetermined map key orders** - see next

 * **Add new kubernetes.UpdateIfChangedSinceLastUpdate method which uses banzaicloud/k8s-objectmatcher to determine whether we need to update an object** - this new method is used in multiple places now where we needed this kind of logic (cluster components, application resources, service provider admin services)

* **Upgrade sigs.k8s.io/application to v0.8.3 (adds ComponentsReady to Status)** - this is needed for checking for no resources (see next)

 * **Check whether the application has any healthy resources** - this fixes an incorrect state check where we considered an application with zero resources as healthy (hence we never realised the autoscaler resource selectors are incorrect)

 * **Recreate the application object when updated** - there is a strange behaviour in the application object, that I couldn't update the resource selectors. I think the app controller might revert them.

 * **Delete the Cluster Autoscaling IAM role if not required or the cluster is deleted** - this wasn't done before

## Testing

What was tested:
 - creating EKS/GKE clusters and deleting them
 - enable/disable autoscaler, use multiple nodegroups with different settings, validate that the autoscaler does its job
 - install AWS servicebroker, delete AWS servicebroker

mrsheepuk commented on pkg/clusterproviders/eks/components.go:
```diff
@@ -94,28 +94,31 @@ func (p Provider) SetComponents(ctx kore.Context, cluster *clustersv1.Cluster, c
+	})
+	if helmOperatorService == nil {
+		return fmt.Errorf("%q service can not be found", helmOperatorName)
+	}
```
we've got this exact logic in two places now (see my earlier comment) both with the same assumption about the operator name - can we share this logic?

mrsheepuk replied:
my mistake, not in two places, but the above comment about hard-coding the assumption about the name applies here.

mrsheepuk commented on pkg/clusterproviders/eks/components.go:
```diff
@@ -66,13 +74,46 @@ func (p Provider) SetComponents(ctx kore.Context, cluster *clustersv1.Cluster, c
+		helmOperatorName := cluster.Name + "-flux-helm-operator"
+		helmOperatorService := components.Find(func(comp kore.ClusterComponent) bool {
+			if service, ok := comp.Object.(*servicesv1.Service); ok {
+				return service.Name == helmOperatorName
```
reviewing commit-by-commit I thought this logic was duplicated, it's not, but the comment here applies to the final version...

bandesz replied:
Added constants for the app names

bandesz commented on pkg/apis/eks/v1alpha1/eksnodegroup_types.go:
```diff
@@ -89,6 +89,9 @@ type EKSNodeGroupStatus struct {
 	NodeIAMRole string `json:"nodeIAMRole,omitempty"`
+	// AutoScalingGroupName is the name of the Auto Scaling Groups belonging to this node group
+	// +listType=set
+	AutoScalingGroupNames []string `json:"autoScalingGroupNames,omitempty"`
```
Sadly the AWS API can return multiple values, I have no clue why and they don't explain it in the API docs (or at least I haven't found any explanation).

bandesz commented on pkg/clusterproviders/eks/components.go:
```diff
@@ -133,10 +185,65 @@ func (p Provider) BeforeComponentsUpdate(ctx kore.Context, cluster *clustersv1.C
+				values := map[string]interface{}{
+					"cloud-provider": "aws",
+					"image": map[string]interface{}{
+						"tag": "v1.16.4",
```
I suspect later we can introduce an autoscaler version parameter maybe, for now I don't have a better way.

bandesz commented on pkg/utils/cloud/aws/eks.go:
```diff
@@ -396,37 +390,29 @@ func (c *Client) UpdateNodeGroup(ctx context.Context, group *eksv1alpha1.EKSNode
-		}); err != nil {
+		input.ScalingConfig.MinSize = aws.Int64(group.Spec.DesiredSize)
+		input.ScalingConfig.MaxSize = aws.Int64(group.Spec.DesiredSize)
+	}
```
I originally removed this whole logic, but then I saw that this is also how it works on the UI and I reverted it. I will add the comments back.

mrsheepuk replied:
Yeah, that logic was hard-won and is well tested, i'd avoid changing it unless you have a strong reason to...

bandesz commented:
@mrsheepuk I think I addressed all comments, changes are here: https://github.com/appvia/kore/compare/d6bb7f677790e2b2c9dfd7508154ef6985389024..7b808b326502821f377573ef4717b5ba35a92366

bandesz commented:
@mrsheepuk I reverted the autoscaling update part completely: https://github.com/appvia/kore/compare/7b808b326502821f377573ef4717b5ba35a92366..e0ce95f056e90ebe2bba4870a314df12f1747e69

bandesz commented:
Found another bug in autoscaling while testing: changing min/max can cause the desired value to be outside of the allowed range. Fix: https://github.com/appvia/kore/pull/1026/commits/40c7da227d87530a89da1855c920fea0ac7747ba
